### PR TITLE
Add unit-test case for #120 and fix generic-import bug of enhanced-for-loop cleanUp

### PR DESCRIPTION
**Pull request by HannesWell**
## What it does

Add a unit test case for issu e#120 with the reproducer presented there.

This test case also shows that the fix for #120, PR #123, falsely adds an additional import for the generic type `V`.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by HannesWell** [comment]
@ktatavarthi or @jjohnstn could you please have a look at this?

**Comment by HannesWell** [comment]
@ktatavarthi or @jjohnstn I think I was able to fix that. Additionally I adjusted some variable names to better fit their value and clarified one code path.
Can you please review this. At the moment the enhanced for-loop conversion is broken.

**Comment by HannesWell** [comment]
@jjohnstn can you please review this PR. If Jeff is not available, @iloveeclipse can you have a look at this?
I would appreciate if the conversion-to-enhanced-forloop gets fixed.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java [hunk 1]
```diff
@@ -418,19 +419,17 @@ public void rewrite(UseIteratorToForLoopFixCore upp, final WhileLoopToChangeHit
 			varBinding= type.resolveBinding();
 		}
 		if (type == null || varBinding == null) {
-			looptargettype= "java.lang.Object"; //$NON-NLS-1$
 			ITypeBinding binding= computeTypeArgument(hit.iteratorDeclaration);
-			Type collectionType= null;
+			Type parametrizedType= null;
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java:423]: Change to parameterizedType.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java [hunk 2]
```diff
@@ -418,19 +419,17 @@ public void rewrite(UseIteratorToForLoopFixCore upp, final WhileLoopToChangeHit
 			varBinding= type.resolveBinding();
 		}
 		if (type == null || varBinding == null) {
-			looptargettype= "java.lang.Object"; //$NON-NLS-1$
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java:421]: Why did you remove this?  This is a fall back type.  In your new code, you set parametrizedType to null.  If the binding is null, you will set the type to null.  Bindings can come back null for many reasons so you need to be prepared.

**Review by jjohnstn — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java (hunk 1 above)
Comment (HannesWell) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java]: Done. Also corrected the type in the handleParametrizedType().

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java (hunk 2 above)
Comment (HannesWell) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java:421]: Sorry, I don't know why, but I must have overseen that `computeTypeArgument()` can return null.
I reverted that part.

**Comment by iloveeclipse** [comment]
I don't know this code at all, so I will not being able to review that.
Beside this, four tests fail, first one:
```
java.lang.ClassCastException: class org.eclipse.jdt.ui.examples.MyQuickAssistProcessor$1 cannot be cast to class org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal (org.eclipse.jdt.ui.examples.MyQuickAssistProcessor$1 is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @206e7912; org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @c042b97)
	at org.eclipse.jdt.ui.tests.quickfix.AssistQuickFixTest1d8.testSurroundWithTryWithResource_01(AssistQuickFixTest1d8.java:5772)
```

**Comment by jjohnstn** [comment]
Yes, those failures are new.  The test runs fine locally.  There is a test that makes its own AssistProcessor which seems to interfere now.  I am going to try changing the tests so they don't assume there is a CUProposal and still work.  I will rebase this change when done.

**Review by jjohnstn — APPROVED** [review]

**Comment by jjohnstn** [comment]
Thanks Hannes.  Merging.

**Comment by HannesWell** [comment]
Thank you Jeff for reviewing.
